### ISSUE TITLE
Incorrect Markdown Line Break Behavior

### ISSUE THREAD

**Issue by mjordan56** [flutter/flutter_markdown#261]
TL; DR - The line break behavior introduce with PR #235, Added support for multiple line breaks, is incorrect Markdown behavior. PR #235 needs to be reverted to the previous behavior to be consistent with the Markdown specification.

@domesticmouse, @lzhuor, @AndroidNils, @grenadier117, and @DaveWoodCom. I'm including all of you in the discussion of this issue because you commented on either issue #221 or #249. Both of these issues are addressing the same line break behavior in the flutter_markdown package.

---

While digging into the parsing code in the flutter_markdown package, I came across this code in widget.dart:

    /// FIXME: Enhance it using a single RegEx
    final List<String> lines = widget.data
        .replaceAll(RegExp(r'[ ]{2}(\r?\n)'), '  &amp;')
        .split(RegExp(r'\r?\n'))
        .map((e) => e?.replaceAll(RegExp(r'  &amp;'), '  \n'))
        .toList();
    final md.Document document = md.Document(
      extensionSet: widget.extensionSet ?? md.ExtensionSet.gitHubFlavored,
      inlineSyntaxes: (widget.inlineSyntaxes ?? [])
        ..add(TaskListSyntax())
        ..map((syntax) => syntax),
      encodeHtml: false,
    );

The "FIXME" tag caught my eye. I wanted to see if I could fix this regular expression and remove the FIXME comment. What I discovered (after lots of time researching hard and soft line breaks in Markdown, the markdown package support for line breaks, and reading the specs for Common Mark and GitHub Flavored Markdown) is that this code is incorrect and goes against the Markdown specs. It's my opinion that the changes in PR #235 should be reverted to the previous behavior which is in accordance with the original Markdown specification, the Common Mark specification, and the GitHub Flavored Markdown specification. I read issue #221, Adding break line (white line) doesn't work, that was the impetus for PR #235. There are a number of incorrect statements and misconceptions about Markdown hard and soft line break behavior in the comments of issue #221. I can address all of the comments in issue #221 as to why the incorrect expected behavior is not consistent with any of the Markdown specifications.

Here are my recommendations to make the flutter_markdown package compliant with correct Markdown line break behavior:

- PR #235 should be reverted
- Fix the line break tests to the correct Markdown behavior
- Improved documentation in the README.md about hard and soft line breaks
- Add a mini-demo to the example suite that demonstrates the line break behaviors

I welcome everyone's comments and technical observations regarding the flutter_markdown package formatting behavior or the Markdown specifications regarding hard and soft line breaking behavior.

Michael Jordan

**Comment by lzhuor** [comment]
Thanks @mjordan56. I am the original author of PR #221 and glad to hear the solution of hard and soft line breaks with flutter_markdown. The original issue is how br tag is parsed. Need to drill into the parsing logic which I did a few months back.

**Comment by mjordan56** [comment]
Hi @lzhuor. Sure, I'd be glad to share what I've discovered regarding Markdown hard and soft line break behavior as defined in the three relevant Markdown reference documents; [John Gruber's Markdown syntax description](https://daringfireball.net/projects/markdown/syntax), [CommonMark specification](https://spec.commonmark.org/0.29/), and [GitHub Flavored Markdown specification](https://github.github.com/gfm/).

The [original Markdown syntax description](https://daringfireball.net/projects/markdown/syntax), often referred to as Markdown.pl as the syntax parser and formatter was originally implemented as a Perl script, was somewhat ambiguous about line breaks. The Markdown syntax overview discusses line breaks in conjunction with paragraph block elements. The specification says,

>PARAGRAPHS AND LINE BREAKS
>
>A paragraph is simply one or more consecutive lines of text, separated by one or more blank lines. (A blank line is any line that looks like a blank line — a line containing nothing but spaces or tabs is considered blank.) Normal paragraphs should not be indented with spaces or tabs.
>
>The implication of the “one or more consecutive lines of text” rule is that Markdown supports “hard-wrapped” text paragraphs. This differs significantly from most other text-to-HTML formatters (including Movable Type’s “Convert Line Breaks” option) which translate every line break character in a paragraph into a \<br /> tag.
>
>When you do want to insert a <br /> break tag using Markdown, you end a line with two or more spaces, then type return.
>
>Yes, this takes a tad more effort to create a \<br />, but a simplistic “every line break is a \<br />” rule wouldn’t work for Markdown. Markdown’s email-style blockquoting and multi-paragraph list items work best — and look better — when you format them with hard breaks.

There are no other references or mention of line break behavior in the Markdown.pl specification. The syntax description describes line break behavior with respect to paragraph blocks. The behavior of hard breaks at the end of a paragraph is ambiguous. The [CommonMark specification notes this ambiguity and the need for a Markdown specification.](https://spec.commonmark.org/0.29/#why-is-a-spec-needed-)

> *NOTE*: The CommonMark and GitHub Flavored Markdown specifications are basically identical on the topic of line breaks. I reference the CommonMark specification in my following comments but the same information is easily found in the GitHub Flavored Markdown specification.

So, what does the [CommonMark specification](https://spec.commonmark.org/0.29/) have to say about line breaks? I believe three sections of the specification are relevant to the Markdown line break behavior we are discussing; [section 4.9 Blank Lines](https://spec.commonmark.org/0.29/#blank-lines), [section 6.9 Hard line breaks](https://spec.commonmark.org/0.29/#hard-line-breaks), and [section 6.10 Soft line breaks](https://spec.commonmark.org/0.29/#soft-line-breaks).

Section 4.9 informs us that "Blank lines between block-level elements are ignored" and "Blank lines at the beginning and end of the document are also ignored."

Section 6.9 informs us that "A line break (not in a code span or HTML tag) that is preceded by two or more spaces and does not occur at the end of a block is parsed as a hard line break" and "For a more visible alternative, a backslash before the line ending may be used instead of two spaces." It also explicitly states "Hard line breaks are for separating inline content within a block. Neither syntax for hard line breaks works at the end of a paragraph or other block element."

And finally, section 6.10 says "A regular line break (not in a code span or HTML tag) that is not preceded by two or more spaces or a backslash is parsed as a softbreak. (A softbreak may be rendered in HTML either as a line ending or as a space. The result will be the same in browsers. In the examples here, a line ending will be used.)"

Given these guidelines, the change in PR #235 makes the formatted output of the Flutter Markdown widget incorrect with the CommonMark specification.

1. Blank lines with two or more spaces are not ignored.
1. Blank lines with two or more spaces at the beginning and end of are not ignored.
1. Hard line breaks *are for separating inline content within a block. Neither syntax for hard line breaks works at the end of a paragraph or other block element.* PR #235 tricks the Markdown parser to incorrectly inject hard line breaks at the end of elements.
1. PR #235 also tricks the Markdown parser to treat white space lines ending in two spaces to be treated like paragraph text elements.
1. PR #235 also introduces a bug where a space character is inserted into the output TextSpan under certain conditions.

This last point is probably the most egregious issue associated with the PR #235 change. The forced injection of newline characters in the lines to be parsed causes the block syntax parser to generate erroneous nodes. For example, consider the following Markdown (spaces are represented by center dots):
```
foo baz・・\n
\n
foo baz
```

The expected CommonMark output should be:
```
foo baz\n
foo baz
```

The expected output I think PR #235 is trying to achieve is:
```
foo baz\n
foo baz
```

But the actual output with PR #235 is:
```
foo baz\n
・foo baz
```

The reason for this behavior is PR #235 forces a hard break at the end of the first line and retains the second line for parsing. The second line's newline is replaced with a space character because Markdown looks to join successive lines in a paragraph and separates them with a space character. It then joins the third line with the second line producing a line that begins with an extraneous and unwanted space character. This is just one example of the manifestation of this problem. I know of at least one more case and believe I've noticed other instances. This problem is hard to see due to the nature of white space and proportional fonts.

So, I content that PR #235 should be reverted for two reasons. First, the implementation is not consistent with the Markdown specification. Second, PR #235 has introduced bugs that produce incorrect output.

This comment is already too long. I've laid out my reasons for reverting PR #235 here. I add a separate comment with my thoughts on how to address the issue of adding vertical white space to the Flutter Markdown widget.

**Comment by lzhuor** [comment]
Thanks @mjordan56 . I am cool to revert #235 and of course the decision is up to the maintainers. 

I only suggest that the multiple break issue should be addressed as well (in a better way I suppose) while we are reverting the changes.

**Comment by mjordan56** [comment]
Hi @lzhuor. I agree there is probably a desire to have a vertical spacing solution available to users of the Markdown widget. I alluded to thinking about solutions to this issue. I have three thoughts on how to provide users of the flutter_markdown package control over the vertical spacing of block elements which translate into Flutter widgets.

Here's what I believe is the issue you and other users are interested in solving, please correct me if I'm wrong.

- The Markdown specification doesn't provide the user with any syntax for vertical positioning control such as inserting blank lines between block elements.
- Traditional Markdown users that target HTML as the output from the Markdown parser and formatter inject raw HTML into their Markdown source documents to add some measure of additional formatting. A common example, which is relevant to this discussion, is the injection of <br /> tags into the Markdown document to force hard line breaks in the output HTML document. This is not a viable option for the Flutter Markdown widget.
- The Markdown widget does support defining a block spacing parameter that is used to define the height of a SizedBox widget that is inserted between block elements but this feature is applied uniformly in the resulting Markdown widget and is not sufficient for all vertical spacing needs.
- Are there other issues that I haven't listed?

My thoughts on how to best address the issue of vertical spacing:

1. Increase the visibility of the block spacing parameter in the README.md documentation. Include a statement that the Wrap Alignment Demo in the example app demonstrates the effect of different block spacing values.
1. Add a cookbook example of how to implement a custom block syntax object and custom builder to create a vertical spacing block element. I created just such a custom block element. I defined the tag `{{sp}}` and an associated builder that inserts a SizedBox widget for the custom block element. I would not include this as part of the flutter_markdown package as this block element is a custom definition and not part of the Markdown specification. I would add this information to the README.md so users that want to implement such behavior could easily see how to do so.
    > *NOTE:* I discovered that the support for custom block syntax and builders appears to be incomplete. For me to add the custom block syntax I described, `{{sp}}`, I had to add code to builder.dart to check for and call my custom element builder. This will need to be fixed to support this solution. @nobuhito implemented the custom syntax and builder changes in PR #241. It would be great if he could comment on the state of his changes.
1. Another idea I've had that I really like is to support decorations for the block elements. This would provide users with a large set of options for customizing the appearance of the widget associated with each block element including the padding of the widget. This is also the solution that is the most expensive as it would require a little work to implement.

**Comment by lzhuor** [comment]
Thanks @mjordan56 , I am not sure `This is not a viable option for the Flutter Markdown widget` is a valid statement. If you check the live demo from Dart markdown, it does support vertical spacing with `\n`. The root cause is that the Flutter Widget does not properly parse the raw Markdown String. To address it we need to figure out how to better recognise the space and pass the block correctly to Dart Markdown. 

Try it here:
https://dart-lang.github.io/markdown/

**Comment by mjordan56** [comment]
@lzhuor I just noticed that the \<br /> element was missing from my comment because it wasn't escaped. I fixed the error.

I contend that this statement is correct. Inserting \<br /> tags into a Markdown document are valid for builders that support HTML output. The markdown package and the Dart Markdown Live Editor are producing HTML output. When the builder for this target comes across HTML tags in the source Markdown document it just preserves the text of the HTML tags and injects them in the output. The markdown package is not supporting any vertical spacing, it's the browser interpreting the HTML that is creating the line break. The markdown package produces HTML that is further processed by a browser. Ther flutter_markdown package creates a widget tree that has completed the construction of widgets.

The output in the HTML pane of the Dart Markdown Live Editor is not valid in this case because the \<br /> is just an HTML tag pass-through for the browser to handle.

**Comment by lzhuor** [comment]
@mjordan56 Sorry, I don't think so. If you debug the code, the Markdown string is intermediately transformed to Node which contains information about the HTML tag. The HTML tag is later used to determine the Widget. I don't have the bandwidth to show the evidence of it. Will probably show it over the weekend as in I need to run the code locally.

**Comment by mjordan56** [comment]
Hi @lzhuor. I'm not sure what point you disagree with. I have debugged both the markdown and flutter_markdown code extensively. I've stepped through the AST creation code and the builder code. I'm intimately familiar with the Visitor pattern used by the builder code to generate the output HTML code by the markdown package and the Flutter widgets generated by the flutter_markdown library. I can help point you to the relevant code associated with line break elements in both the markdown and flutter_markdown packages.

Markdown output is treated as a column of block elements. The first thing to note in the markdown package is that there is no BlockSyntax for handling line breaks. If you then look at the html_renderer.dart file line 42 you'll see the list of HTML tags supported in the `_blockTags` list. The HTML tag \<br /> is not in the list of generated HTML tags. If we now look at the inline_parser.dart, this is where we find the LineBreakSyntax class. This is the code that inserts a 'br' element into the AST. Note that this is an inline element and not a block element so it has to be within the confines of a block element parent. So what happens is the block parsing is performed first to generate all the block elements. All lines of text are contained in an appropriate block element parent. When building the block elements the document parser will join adjacent lines under some circumstances, such as lines in a paragraph. It's in these cases if there is a line of text in a block element the inline parser will recognize the '\\' or multiple spaces at the end of a line and trigger the LineBreakSyntax object to create a 'br' element in the AST. Either of these character sequences on the last line of the block element will not match the LineBreakSyntax because the `node.textContent` property string doesn't end with a '\\n' character.

The flutter_markdown builder.dart code implements the Visitor pattern to traverse the AST and generate appropriate Flutter widgets from the elements in the tree. It is only interpreting the structure to build an appropriate widget tree. It does not create line break elements.

So, I stand by my observations. A '\<br />' tag inserted into a Markdown document is ignored by the document parser, it does not generate an element in the AST, and is just passed through to the target browser in the HTML to interpret and render.

In addition, I still believe the correct behavior, regardless of how the markdown package behaves, is to follow the Markdown syntax specification. I see no discussion of line break or vertical spacing behavior in any of the documents. The behavior of the markdown package is consistent with the Markdown syntax specifications.

BTW, when I debugged the flutter_markdown and markdown package with PR #235 implemented the document parser did not convert the additional '\\n' newline characters into 'br' elements in the AST. The newline character remained in the text of the nodes and was processed by the inline LineBreakSyntax object. The problem with this approach is that there is other code that doesn't expect newline characters to be present hence the bugs introduce like the one I mentioned previously where an unexpected and unwanted blank space will appear at the beginning of a line of text.

**Comment by mjordan56** [comment]
@lzhuor I just submitted PR #264 to revert the changes made by PR #235. I'm going to ask @domesticmouse to please review and merge my PR.

**Comment by domesticmouse** [comment]
I've merged #264. I'm curious what work remains to be done before this issue can also be closed out?

**Comment by mjordan56** [comment]
@domesticmouse Now that PR #264 has been merged this issue can be closed. There is some discussion about a suitable technique for injecting new lines or vertical spacing into the Markdown widget but that should be a separate issue. Thanks!

**Comment by domesticmouse** [comment]
Closing. Thanks all!
